File: sdk/derive/blockid.go
package derive

import (
	"encoding/hex"
	"strconv"

	"github.com/bpprotocol/blockparty/sdk/crypto"
)

// GetBlockID derives the content-binding identifier for a block:
//
//	dataHash = hex(Keccak256(data))
//	message  = "block.v"+version+":"+ts+"/"+address+"/"+typeCode+"/"+dataHash
//	id       = hex(SHA256(HMAC_SHA256(key=audienceCode, message)))
//
// Including the payload hash makes the ID change whenever data changes, so the
// ID commits to the block's contents.
func GetBlockID(version uint32, timestamp int64, audienceCode Code, address Address, typeCode Code, data []byte) string {
	dataHash := hex.EncodeToString(crypto.Keccak256(data))
	message := "block.v" + strconv.FormatUint(uint64(version), 10) +
		":" + strconv.FormatInt(timestamp, 10) +
		"/" + string(address) +
		"/" + typeCode.Hex() +
		"/" + dataHash
	seed := crypto.HMACSHA256([]byte(audienceCode.Hex()), []byte(message))
	return hex.EncodeToString(crypto.Sha256(seed))
}

File: js/packages/sdk/src/derive/blockid.ts
import { hmacSha256, keccak256, sha256 } from "../crypto/index.js";
import { bytesToHex, utf8ToBytes } from "../bytes.js";

/**
 * Derive the content-binding block identifier:
 *
 *   dataHash = hex(keccak256(data))
 *   message  = "block.v"+version+":"+timestamp+"/"+address+"/"+typeCodeHex+"/"+dataHash
 *   id       = hex(sha256(hmacSha256(key = hex(audienceCode), message)))
 *
 * Including the payload hash makes the ID change whenever data changes.
 * `timestamp` is unix seconds; values stay well within Number's safe range.
 */
export function getBlockID(
  version: number,
  timestamp: number,
  audienceCode: Uint8Array,
  address: string,
  typeCode: Uint8Array,
  data: Uint8Array,
): string {
  const dataHash = bytesToHex(keccak256(data));
  const message =
    "block.v" +
    version +
    ":" +
    timestamp +
    "/" +
    address +
    "/" +
    bytesToHex(typeCode) +
    "/" +
    dataHash;
  const seed = hmacSha256(utf8ToBytes(bytesToHex(audienceCode)), utf8ToBytes(message));
  return bytesToHex(sha256(seed));
}
